<a href="https://colab.research.google.com/github/aanchalkhanna72/aanchalkhanna72/blob/main/E22001_Study_2_AML.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

**E22001 Aanchal Khanna, 05/06/2022**

**Objective: To calculate coefficient estimates of Multiple Linear Regression**

In [2]:
#importing libraries
import numpy as np
import pandas as pd
import statsmodels.api as sm

/usr/local/lib/python3.7/dist-packages/statsmodels/tools/_testing.py:19: FutureWarning: pandas.util.testing is deprecated. Use the functions in the public API at pandas.testing instead.
  import pandas.util.testing as tm


In [3]:
#importing dataset
cars=pd.read_csv("https://raw.githubusercontent.com/aanchalkhanna72/aanchalkhanna72/main/cars.csv",sep=",")

In [4]:
#reading top 5 rows of dataset
cars.head()

Car   MPG  Cylinders  Displacement  Horsepower  \
0  Chevrolet Chevelle Malibu  18.0          8         307.0         130   
1          Buick Skylark 320  15.0          8         350.0         165   
2         Plymouth Satellite  18.0          8         318.0         150   
3              AMC Rebel SST  16.0          8         304.0         150   
4                Ford Torino  17.0          8         302.0         140   

   Weight  Acceleration  Model Origin  \
0    3504          12.0     70     US   
1    3693          11.5     70     US   
2    3436          11.0     70     US   
3    3433          12.0     70     US   
4    3449          10.5     70     US   

                                          Unnamed: 9 Unnamed: 10 Unnamed: 11  \
0                                                NaN         NaN         NaN   
1                                                NaN         NaN         NaN   
2                                          For every    entity,    there are   
3  each variable represents a characteristic of  ...         NaN         NaN   
4                                  entity - 406 cars         NaN         NaN   

                         Unnamed: 12  
0                                NaN  
1                                NaN  
2  7 values so this is 1 observation  
3                                NaN  
4                                NaN

In [5]:
#choosing columns of interest for multiple regression
X=cars[["Weight","Horsepower","Displacement"]]
X=sm.add_constant(X)

/usr/local/lib/python3.7/dist-packages/statsmodels/tsa/tsatools.py:117: FutureWarning: In a future version of pandas all arguments of concat except for the argument 'objs' will be keyword-only
  x = pd.concat(x[::order], 1)


In [6]:
X

const  Weight  Horsepower  Displacement
0      1.0    3504         130         307.0
1      1.0    3693         165         350.0
2      1.0    3436         150         318.0
3      1.0    3433         150         304.0
4      1.0    3449         140         302.0
..     ...     ...         ...           ...
401    1.0    2790          86         140.0
402    1.0    2130          52          97.0
403    1.0    2295          84         135.0
404    1.0    2625          79         120.0
405    1.0    2720          82         119.0

[406 rows x 4 columns]

In [7]:
y=cars.MPG
y

0      18.0
1      15.0
2      18.0
3      16.0
4      17.0
       ... 
401    27.0
402    44.0
403    32.0
404    28.0
405    31.0
Name: MPG, Length: 406, dtype: float64

In [8]:
#top 5 rows of X
X.head()

const  Weight  Horsepower  Displacement
0    1.0    3504         130         307.0
1    1.0    3693         165         350.0
2    1.0    3436         150         318.0
3    1.0    3433         150         304.0
4    1.0    3449         140         302.0

In [1]:
#Fitting the model is equivalent to estimating  the coefficients of features i.e. b0,b1,b2,b3
#Coefficients can be expressed as:

Multiple Linear Regression can be represented as:

yi=b0+b1*x1i+b2*x2i+b3*x3i+b4*x4i

y=b0+b1*(Weight)+b2*(Horsepower)+b3*(Displacement)+Ei, i=1(1)n


In [24]:
#converting dataframes to arrays
X_arr=X.to_numpy()
y_arr=y.to_numpy()

In [30]:
#using the derived formula  for calculatinng b curl, whose array will give us resepctive coefficient estimates of Weight, Horsepower, Displacement
((np.linalg.inv(X_arr.T@X))@X_arr.T)@y_arr



array([ 4.44999776e+01, -5.31300288e-03, -3.93767539e-02, -7.91897363e-03])

**According to our calculations, the coefficient estimates are as follows:
constant: 44.499, Weight: -0.0053, Horsepower: -0.0393, Displacement: -0.0079**

In [27]:
#confirming answers through statsmodel method
model=sm.OLS(y,X).fit()
model.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                    MPG   R-squared:                       0.635
Model:                            OLS   Adj. R-squared:                  0.633
Method:                 Least Squares   F-statistic:                     233.5
Date:                Sun, 05 Jun 2022   Prob (F-statistic):           1.08e-87
Time:                        09:27:26   Log-Likelihood:                -1234.9
No. Observations:                 406   AIC:                             2478.
Df Residuals:                     402   BIC:                             2494.
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
================================================================================
                   coef    std err          t      P>|t|      [0.025      0.975]
--------------------------------------------------------------------------------
const           44.5000      1.394     31.929      0.000      41.760      47.240
Weight          -0.0053      0.001     -6.354      0.000      -0.007      -0.004
Horsepower      -0.0394      0.013     -3.033      0.003      -0.065      -0.014
Displacement    -0.0079      0.008     -1.054      0.293      -0.023       0.007
==============================================================================
Omnibus:                       98.313   Durbin-Watson:                   1.039
Prob(Omnibus):                  0.000   Jarque-Bera (JB):              636.332
Skew:                          -0.849   Prob(JB):                    6.64e-139
Kurtosis:                       8.893   Cond. No.                     1.71e+04
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 1.71e+04. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

In [29]:
model.params

const           44.499978
Weight          -0.005313
Horsepower      -0.039377
Displacement    -0.007919
dtype: float64

**According to the statsmodel method, the coefficient estimates are as follows:
constant: 44.5, Weight: -0.0053, Horsepower: -0.0393, Displacement: -0.0079. We notice that the results are the same.**

###Interpretation of the estimates

*   Weight: Keeping all else equal, as the weight increases by 1000 pounds, the MPG of a car decreases by 5.3 miles.
*   HP: Keeping all else equal, as the power of an engine increasess by 100 horsepowers, the MPG of a cars decreases by 3.93 miles.
*   Displacement: Keeping all else equal, as the engine displacement increases by 1000 cubic centimetres (cc), MPG of a  car decreases by 7.92 miles.